src/library_pthread.js

```diff
@@ -1050,40 +1050,6 @@ var LibraryPThread = {
     throw 'Atomics.wake returned an unexpected value ' + ret;
   },
 
-  // Returns the number of threads (>= 0) woken up, or one of the values -EINVAL or -EAGAIN on error.
-  emscripten_futex_wake_or_requeue__deps: ['_main_thread_futex_wait_address'],
-  emscripten_futex_wake_or_requeue: function(addr, count, addr2, cmpValue) {
-    if (addr <= 0 || addr2 <= 0 || addr >= HEAP8.length || addr2 >= HEAP8.length || count < 0
-      || addr&3 != 0 || addr2&3 != 0) {
-      return -{{{ cDefine('EINVAL') }}};
-    }
-
-    // See if main thread is waiting on this address? If so, wake it up by resetting its wake location to zero,
-    // or move it to wait on addr2. Note that this is not a fair procedure, since we always wake main thread first before
-    // any workers, so this scheme does not adhere to real queue-based waiting.
-    var mainThreadWaitAddress = Atomics.load(HEAP32, __main_thread_futex_wait_address >> 2);
-    var mainThreadWoken = 0;
-    if (mainThreadWaitAddress == addr) {
-      // Check cmpValue precondition before taking any action.
-      var val1 = Atomics.load(HEAP32, addr >> 2);
-      if (val1 != cmpValue) return -{{{ cDefine('EAGAIN') }}};
-
-      // If we are actually waking any waiters, then new main thread wait location is reset, otherwise requeue it to wait on addr2.
-      var newMainThreadWaitAddress = (count > 0) ? 0 : addr2;
-      var loadedAddr = Atomics.compareExchange(HEAP32, __main_thread_futex_wait_address >> 2, mainThreadWaitAddress, newMainThreadWaitAddress);
-      if (loadedAddr == mainThreadWaitAddress && count > 0) {
-        --count; // Main thread was woken, so one less workers to wake up.
-        mainThreadWoken = 1;
-      }
-    }
-
-    // Wake any workers waiting on this address.
-    var ret = Atomics.wakeOrRequeue(HEAP32, addr >> 2, count, addr2 >> 2, cmpValue);
-    if (ret == Atomics.NOTEQUAL) return -{{{ cDefine('EAGAIN') }}};
-    if (ret >= 0) return ret + mainThreadWoken;
-    throw 'Atomics.wakeOrRequeue returned an unexpected value ' + ret;
-  },
-
```

Comment:
when i compile with:
`emcc -s USE_PTHREADS=1 tests/hello_world.c`

i get the following error :

```
ReferenceError: Atomics is not defined
    at Object.initMainThreadBlock (D:\WORK\ue4\gh_emscripten\a.out.js:1834:9)
    at Object.<anonymous> (D:\WORK\ue4\gh_emscripten\a.out.js:6639:38)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
```
is there another flag i should be using?

Reply:
also, i have merged jukka's multi-threaded stuff with emscripten 1.38.1 -- you can see this here:
https://github.com/nickshinpho/emscripten/commit/d4581636e087a37f316a447510cd4d002cd649ee

@juj @kripken  -- i would like to start a thread on trying to get the multi-threading stuff working with the **UE4 engine** on **FF**.  we are also going to be on-boarding people from **chrome** to also start working on getting their browser with wasm supporting multithreading.  but, due to their security policy -- they will not be able to work within Epic's VPN -- so i would like to see if we can all get this done on **github**. ^_^

*please let me know what the proper etiquette in how this is done.*  i have not used github on such a large project like this (e.g. we use slack and email at epic...) i.e. is it ok for me to just poke this comment in here like this?  not sure how the follow up is handled from random spots.

for now,  the error above is a little strange and i would like to know what flags or switches i need to fix that error.  thanks!

Reply:
About that node error, it's possible node.js does not have atomics, or has it behind a flag. My guess would be v8 still has it behind a flag, so that would need to be passed to node. But anyhow - I don't think multithreading is expected to work in the shell (do workers exist there? setTimeout? I'm not sure, but I'd guess no), why are you trying to run it there?

About etiquette: No problem! :) Feel free to comment in relevant places like here of course. If it's more of a general question and not specific to a PR, feel free to open a new issue with that. We may also want to open a general "multithreading UE4" issue, if that's helpful for you.

Reply:
thanks for the notes.

(just to close my comments) so, i recompiled with `-o hello.html` and tried running it in FF:
- version 60.0.1 got: `ReferenceError: Atomics is not defined`
- version 62.0a1 got: `shared memory is disabled`

seems the `missing Atomic` error is fixed with FF nightly.  and the `shared memory` issue is mention here: https://github.com/kripken/emscripten/issues/6119 -- which, i'll follow that issue page...

thanks again!